op-node/rollup/derive/batches.go

```diff
@@ -1,106 +1,121 @@
-					"epoch", batch.Epoch(), "timestamp", batch.Timestamp, "txs", len(batch.Transactions))
-			} else {
-				log.Warn("filtered batch", "expected_epoch", epoch, "min", minL2Time, "max", maxL2Time,
-					"epoch", batch.Epoch(), "timestamp", batch.Timestamp, "txs", len(batch.Transactions), "err", err)
-			}
-			continue
-		}
-		// Check if we have already seen a batch for this L2 block
-		if _, ok := uniqueTime[batch.Timestamp]; ok {
-			log.Warn("duplicate batch", "epoch", batch.Epoch(), "timestamp", batch.Timestamp, "txs", len(batch.Transactions))
-			// block already exists, batch is duplicate (first batch persists, others are ignored)
-			continue
-		}
-		uniqueTime[batch.Timestamp] = struct{}{}
-		out = append(out, batch)
+type BatchValidity uint8
+
+const (
+	// BatchDrop indicates that the batch is invalid, and will always be in the future, unless we reorg
+	BatchDrop = iota
+	// BatchAccept indicates that the batch is valid and should be processed
+	BatchAccept
+	// BatchUndecided indicates we are lacking L1 information until we can proceed batch filtering
+	BatchUndecided
+	// BatchFuture indicates that the batch may be valid, but cannot be processed yet and should be checked again later
+	BatchFuture
+)
+
+func CheckBatch(cfg *rollup.Config, log log.Logger, l1Blocks []eth.L1BlockRef, l2SafeHead eth.L2BlockRef, batch *BatchWithL1InclusionBlock) BatchValidity {
+	nextTimestamp := l2SafeHead.Time + cfg.BlockTime
+	if batch.Batch.Timestamp > nextTimestamp {
+		// no log, very common case, this happens when batches are included early or out of order.
+		return BatchFuture
 	}
-	return
-}
+	// add details to the log
+	log = log.New(
+		"batch_timestamp", batch.Batch.Timestamp,
+		"parent_hash", batch.Batch.ParentHash,
+		"batch_epoch", batch.Batch.Epoch(),
+		"txs", len(batch.Batch.Transactions),
+	)
 
-func ValidBatch(batch *BatchData, config *rollup.Config, epoch eth.BlockID, minL2Time uint64, maxL2Time uint64) error {
-	if batch.EpochNum != rollup.Epoch(epoch.Number) {
-		// Batch was tagged for past or future epoch,
-		// i.e. it was included too late or depends on the given L1 block to be processed first.
-		// This is a very common error, batches may just be buffered for a later epoch.
-		return DifferentEpoch
+	// sanity check we have consistent inputs
+	if len(l1Blocks) == 0 {
+		log.Warn("missing L1 block input, cannot proceed with batch checking")
+		return BatchUndecided
 	}
-	if batch.EpochHash != epoch.Hash {
-		return fmt.Errorf("batch was meant for alternative L1 chain")
+	if l1Blocks[0].Hash != l2SafeHead.L1Origin.Hash {
+		log.Warn("safe L2 head L1 origin does not match batch first l1 block",
+			"safe_l2", l2SafeHead, "safe_origin", l2SafeHead.L1Origin, "l1_block", l1Blocks[0])
+		return BatchUndecided
 	}
-	if (batch.Timestamp-config.Genesis.L2Time)%config.BlockTime != 0 {
-		return fmt.Errorf("bad timestamp %d, not a multiple of the block time", batch.Timestamp)
+
+	if batch.Batch.Timestamp < nextTimestamp {
```

Comment:
this plus `batch.Batch.Timestamp > nextTimestamp` is a weird way of saying
```
if batch.Batch.Timestamp != nextTimestamp {
    return ...
}
```

Reply:
True, although the results are different: too old = drop, too new = buffer, exact equal = process.